### PR TITLE
feat(grpo): add Perfetto workload tracing

### PR DESCRIPTION
**Pull request by youngeunkwon0405**
## Summary

- extend the opt-in Perfetto workload trace from #3879 to GRPO
- trace the driver timeline for legacy synchronous GRPO, TransferQueue synchronous GRPO, and asynchronous GRPO
- merge async collector and sync rollout-actor events into the driver trace
- add native per-sample rollout tracks when the selected generation backend uses the asynchronous rollout path
- document PPO and GRPO configuration and trace lanes together

This PR is stacked on #3879. Please review only the GRPO-specific commit/diff; the shared tracer and PPO support remain in the base PR.

## Usage

```yaml
logger:
  perfetto:
    enable: true
    name: grpo_perfetto.json
```

The same options can be set from the CLI:

```bash
uv run examples/run_grpo.py --config <recipe.yaml> \
  logger.perfetto.enable=true \
  logger.perfetto.name=/shared/path/grpo_perfetto.json
```

Open the resulting Chrome trace-event JSON at https://ui.perfetto.dev/. The trace includes the driver schedule, nested timer spans, step markers, async replay-buffer metadata, collector/rollout-actor activity, and native sample tracks where supported by the rollout backend.

## Example traces

NVIDIA access is required for these shared files:

- [Synchronous GRPO trace](https://drive.google.com/file/d/17XB3VjOlCVqgngmHd2IDhYbeEGmHU7Yl/view?usp=drive_link)
- [Asynchronous GRPO trace](https://drive.google.com/file/d/1hNYt1fDFBQACSjUE-OjVTyeHY1uMAI16/view?usp=drive_link)

## Validation

- CW-DFW job `16766384`: Ruff format/check clean, full-project Pyrefly `0 errors`, and `779 passed, 6 skipped`
- CW-DFW sync job `16767609`: `COMPLETED (0:0)`, two GRPO steps, 140 trace events, 8 sample tracks
- CW-DFW async job `16769381`: `COMPLETED (0:0)`, two GRPO steps, replay-buffer trajectory ages 0.0 and 1.0, 205 trace events, 16 sample tracks, collector process merged
- both JSON files were parsed programmatically; timestamps/durations are finite, durations are non-negative, and expected driver/step/sample metadata is present

## Scope

Per-sample tracks require the native asynchronous rollout implementation. Batched synchronous and NeMo Gym rollout backends still appear as driver-side generation spans, but do not expose independent sample timing.

### CODE REVIEW

File: nemo_rl/algorithms/grpo.py [hunk 1]
```diff
@@ -2956,6 +2980,15 @@ def grpo_train(
                     flush=True,
                 )
 
+            driver_trace.instant(
```
Comment (youngeunkwon0405) [inline nemo_rl/algorithms/grpo.py:3046]: [`nemo_rl/algorithms/grpo.py:2983`](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L2983)

**[BUG]** Under dynamic sampling, the `continue` at [grpo.py:3272](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L3272) re-enters the loop top without incrementing `total_steps`, so one training step emits one `grpo_step_start` instant and one `grpo_training_step` counter sample per generation batch (all with the same `step`), while `grpo_step_complete` fires once. The `trace_prefix=f"step {N}"` at [grpo.py:3161](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L3161) is also reused across those gen batches, so sample tracks carry sequential spans from different generation rounds (harmless — sequential span-id reuse doesn't raise — just ambiguous to read).

`dynamic_sampling_num_gen_batches` is `0` exactly on the first gen batch at this emission site, so one option is to guard on it (and the same applies to [grpo_sync.py:603](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo_sync.py#L603)); alternatively, including the gen-batch counter in the args/prefix (e.g. `f"step {total_steps + 1} / gen {dynamic_sampling_num_gen_batches}"`) makes the repeats self-describing.

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -4589,6 +4669,7 @@ def async_grpo_train(
             # exits right after this return.
             checkpointer.shutdown()
             try:
+                _save_async_trace()
```
Comment (youngeunkwon0405) [inline nemo_rl/algorithms/grpo.py:4935]: [`nemo_rl/algorithms/grpo.py:4672`](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4672)

**[BUG]** Two trace-loss gaps in this PR's new instrumentation (same theme as the comment on #3879, but these live in this PR's files and need their own edit):

1. The "dataloader exhausted while waiting for initial buffer fill" `RuntimeError` at [grpo.py:4754](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4754) sits *before* the `try:` at [grpo.py:4770](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4770) whose `finally` saves the trace — yet by this point the collector has been actively generating rollouts (started at [grpo.py:4607](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4607)), so its trace is exactly the diagnostic you'd want for this failure. The sibling pre-loop failures at [4588](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4588)/[4600](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4600)/[4672](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4672) all save. Calling `_save_async_trace()` before that raise (or hoisting the `try` to cover the buffer wait) closes the gap.
2. `grpo_train` and `grpo_train_sync` save only at explicit returns with no `try/finally` around their training loops ([grpo.py:2841](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L2841), [grpo_sync.py:420](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo_sync.py#L420)), so an unhandled mid-run exception loses the whole trace — while `async_grpo_train` correctly saves in its `finally` ([grpo.py:5743](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L5743)). Wrapping both sync loops in `try: ... finally: _save_sync_trace()` would match the async behavior and collapse the explicit call sites.

File: nemo_rl/algorithms/grpo_sync.py [hunk 3]
```diff
@@ -404,7 +405,35 @@ def grpo_train_sync(
     Parity with the legacy path is verified by running the same config
     against both entrypoints and diffing the wandb runs.
     """
-    timer = Timer()
+    trace_enabled, trace_output_path = resolve_trace_config(master_config.logger)
+    driver_trace = Tracer(
+        "driver",
+        virtual_process_name="grpo_sync_driver_events",
```
Comment (youngeunkwon0405) [inline nemo_rl/algorithms/grpo_sync.py]: [`nemo_rl/algorithms/grpo_sync.py:411`](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo_sync.py#L411)

**[GUIDELINE]** Neither this driver nor `async_grpo_train`'s ([grpo.py:4325](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L4325), `"grpo_async_driver_events"`) ever calls `start_async_span` — the TQ rollout lanes come from the actor's own tracer and the async lanes from the collector's — but virtual-process metadata is emitted eagerly at construction, so `grpo_sync_driver_events` and `grpo_async_driver_events` appear as permanently empty process groups in the Perfetto UI (and neither name is in the docs). Same pattern as flagged on #3879 for `ppo_async_driver_events`; for these two constructions, dropping `virtual_process_name` keeps the trace layout matching the docs. (Legacy `grpo_train`'s `grpo_sync_rollouts` at [grpo.py:2831](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/algorithms/grpo.py#L2831) genuinely receives sample tracks and is fine.)

File: docs/perfetto-tracing.md [hunk 4]
```diff
@@ -37,40 +39,48 @@ This also makes command-line enablement available without adding new keys:
...
 - **trajectory_collector_actor**: asynchronous collector work and wait states.
-- **ppo_sync_rollouts** or **ppo_async_rollouts**: one virtual track per native
-  rollout sample. Each track contains the full sample rollout, turns,
-  generation calls, reward calculation, and environment tokenization.
-
-`ppo_step_start`, `ppo_epoch_start`, and `ppo_step_complete` events carry the
-step and PPO-epoch numbers. Async step-complete events also include the replay
-buffer size and average trajectory age. A sample's outer span includes token
-counts, termination/truncation state, and total reward.
+- **sync_rollout_actor**: the TransferQueue GRPO rollout actor, when that sync
```
Comment (youngeunkwon0405) [inline docs/perfetto-tracing.md:62]: [`docs/perfetto-tracing.md:62`](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/docs/perfetto-tracing.md#L62)

**[DOC]** This bullet describes the `sync_rollout_actor` group as showing the TransferQueue rollout actor's work, but the actor's tracer is only ever used as the async-span sink ([sync_rollout_actor.py:286](https://github.com/NVIDIA-NeMo/RL/blob/d263a1803001d29ab2308bad9352727ce6af8538/nemo_rl/experience/sync_rollout_actor.py#L286)) — those spans land on the virtual `grpo_sync_rollouts` group, and no span/instant/timer is wired to the actor's regular lane (contrast the trajectory collector, which wires `ThreadSafeTimer(trace=self._trace)`). As written, the documented group contains only metadata, and is empty on the batched/NeMo-Gym paths too. Consider rewording the bullet around `grpo_sync_rollouts`, or wiring `rollout_to_tq`'s dispatch into a span so the group has content.

**Review by youngeunkwon0405 — COMMENTED** [review]
Multi-agent review pass (RL-codebase expert, test reviewer, independent bug scan, adversarial verification; ruff clean). Nice follow-up — this closes the GRPO trace-harvesting gap flagged on #3879, the rollout-dispatch restructure in `sync_rollout_actor` preserves semantics exactly, the async shutdown save ordering (harvest before `ray.kill`) is correct, and all GRPO exemplar configs inherit the new block via `defaults:`, so the two updated YAMLs are the only ones needed. No new tests warranted — the base PR's `test_trace.py` covers the shared machinery and the reference-config guard passes.

Inline comments cover: duplicate step markers under dynamic sampling, two trace-loss gaps (sync paths lack a `finally`; one async pre-`try` failure path), two empty virtual process groups, and one docs bullet describing a group that never receives spans. All low severity.

_Generated by Claude Code_
